kong/clustering/init.lua

```diff
@@ -5,67 +5,116 @@ local pl_file = require("pl.file")
 local pl_tablex = require("pl.tablex")
 local ssl = require("ngx.ssl")
 local openssl_x509 = require("resty.openssl.x509")
+local isempty = require("table.isempty")
+local isarray = require("table.isarray")
+local nkeys = require("table.nkeys")
+local new_tab = require("table.new")
 local ngx_null = ngx.null
 local ngx_md5 = ngx.md5
+local ngx_md5_bin = ngx.md5_bin
 local tostring = tostring
 local assert = assert
 local error = error
 local concat = table.concat
+local pairs = pairs
 local sort = table.sort
 local type = type
+local min = math.min
 
 
 local MT = { __index = _M, }
 
 
-local compare_sorted_strings
-
-
 local function to_sorted_string(value)
   if value == ngx_null then
     return "/null/"
   end
 
   local t = type(value)
-  if t == "table" then
-    local i = 1
-    local o = { "{" }
-    for k, v in pl_tablex.sort(value, compare_sorted_strings) do
-      o[i+1] = to_sorted_string(k)
-      o[i+2] = ":"
-      o[i+3] = to_sorted_string(v)
-      o[i+4] = ";"
-      i=i+4
-    end
-    if i == 1 then
-      i = i + 1
-    end
-    o[i] = "}"
-
-    return concat(o, nil, 1, i)
-
-  elseif t == "string" then
+  if t == "string" then
     return "$" .. value .. "$"
 
   elseif t == "number" then
-    return "#" .. tostring(value) .. "#"
+    return "#" .. value .. "#"
 
   elseif t == "boolean" then
     return "?" .. tostring(value) .. "?"
 
+  elseif t == "table" then
+    if isempty(value) then
+      return "{}"
+
+    elseif isarray(value) then
+      local count = #value
+      if count == 1 then
+        return "{" .. to_sorted_string(value[1]) .. "}"
+
+      elseif count == 2 then
+        return "{" .. to_sorted_string(value[1]) .. ";"
+                   .. to_sorted_string(value[2]) .. "}"
+
+      elseif count == 3 then
+        return "{" .. to_sorted_string(value[1]) .. ";"
+                   .. to_sorted_string(value[2]) .. ";"
+                   .. to_sorted_string(value[3]) .. "}"
+
+      elseif count == 4 then
+        return "{" .. to_sorted_string(value[1]) .. ";"
+                   .. to_sorted_string(value[2]) .. ";"
+                   .. to_sorted_string(value[3]) .. ";"
+                   .. to_sorted_string(value[4]) .. "}"
+
+      elseif count == 5 then
+        return "{" .. to_sorted_string(value[1]) .. ";"
+                   .. to_sorted_string(value[2]) .. ";"
+                   .. to_sorted_string(value[3]) .. ";"
+                   .. to_sorted_string(value[4]) .. ";"
+                   .. to_sorted_string(value[5]) .. "}"
```

Comment:
Could you elaborate on this block? This looks arbitrarily restrictive (and redundant) to me.

Also, the unittests do not include a case for this whole section.

Reply:
The whole idea of this function is just to get MD5 hash out of the (arbitrary) lua table. The function is rather generic currently and further optimizations can be done with being opinionated, e.g. being opinionated that all table (not array) keys are strings (like they are in db export generated tables currently). The problem of using something like cjson.encode or the upcoming: https://github.com/LuaJIT/LuaJIT/blob/v2.1/doc/ext_buffer.html#L89-L90 is that they can be undeterministic. At least the cjson.encode is. Because `pairs` over Lua table is undeterministic too.

This code here is just to skip the table creation and md5 on small arrays. We have a lot of small arrays in our db exports. It is a hand optimization for this particular case.

Reply:
ok, got it. thanks for clarifying.

Reply:
There should be now unit test that cover this section.